src/overlays/actors/ovl_En_Encount4/z_en_encount4.c

```diff
@@ -33,20 +32,171 @@ const ActorInit En_Encount4_InitVars = {
+                Actor_MarkForDeath(&this->actor);
+            } else if ((this->unk148 != 0) || (this->actor.xzDistToPlayer < 240.0f)) {
+                if ((this->unk148 == 0) && (stalchild->limbCount != 0)) {
+                    Actor_MarkForDeath(&this->actor);
+                } else {
+                    params = 0;
+                    if ((this->unk148 == 0) || (stalchild->limbCount != 0)) {
+                        i = 0;
+                        if (this->unk148 != 0) {
+                            params = 1;
+                            i = 2;
+                        }
+                        while (i < ARRAY_COUNT(D_809C46DC)) {
+                            rotY = D_809C46D0[i] + this->actor.world.rot.y;
+                            pos.x = Math_SinS(rotY) * D_809C46DC[i] + this->actor.world.pos.x;
+                            pos.y = this->actor.world.pos.y;
+                            pos.z = Math_CosS(rotY) * D_809C46DC[i] + this->actor.world.pos.z;
+                            Actor_SpawnAsChild(&play->actorCtx, &this->actor, play, ACTOR_BG_FIRE_WALL, pos.x, pos.y,
+                                               pos.z, 0, this->actor.world.rot.y, 0, params);
+                            i++;
+                        }
+                        this->actionFunc = func_809C42A8;
+                    }
+                }
+            }
+        }
+    }
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_En_Encount4/func_809C3FD8.s")
+void func_809C42A8(EnEncount4* this, PlayState* play) {
+    Player* actor = GET_PLAYER(play);
+    EnSkb* stalchild = this->stalchild;
+    Vec3f pos;
+    f32 yIntersect;
+    s16 yRot;
+    CollisionPoly* colPoly;
+    s32 bgId;
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_En_Encount4/func_809C4078.s")
+    if (this->switchFlags >= 0) {
+        if (Flags_GetSwitch(play, this->switchFlags)) {
+            this->unk150 = 100;
+            this->actionFunc = func_809C464C;
+            return;
+        }
+    }
+    if (this->unk148 == 1) {
+        if ((this->stalchild->actor.id != ACTOR_EN_BSB) || (stalchild->actor.update == NULL)) {
+            Actor_MarkForDeath(&this->actor);
+        }
+    } else if (this->unk14E >= 2) {
+        this->unk150 = 100;
+        this->actionFunc = func_809C464C;
+    } else if (!(gSaveContext.save.weekEventReg[85] & 0x40) && (this->unk14C < 2) &&
+               !(this->actor.xzDistToPlayer > 240.0f)) {
+        pos.x = (Math_SinS(this->actor.world.rot.y) * 30.0f) + this->actor.world.pos.x;
+        pos.y = actor->actor.floorHeight + 120.0f;
+        pos.z = (Math_CosS(this->actor.world.rot.y) * 30.0f) + this->actor.world.pos.z;
+        yIntersect = BgCheck_EntityRaycastFloor5(&play->colCtx, &colPoly, &bgId, &this->actor, &pos);
+        if (!(yIntersect <= -32000.0f) && ((actor->actor.depthInWater == -32000.0f ||
+                                            !(yIntersect < (actor->actor.world.pos.y - actor->actor.depthInWater))))) {
```

Comment:
The inverted way all this is written makes me wonder if it might be better to just use an early return (it does match)